File: coins-on-the-clock/python/coins-on-the-clock.py
import itertools
import time

'''
_GetValidSequences
Gets list of list of ints
Each list of ints represents a valid coin sequence on the clock
numHours        - Int. Number of hours on the clock
coins           - Array of ints. Must be matched with counts. Gives value of each coin
counts          - Array of ints. Must be matched with coins. Gives count of each coin
clockState      - Array of bools. Lists which clock hours have coins
returnValues    - Reference to a list to push solutions onto
currentSequence - List of ints. Current sequence of coins placed
currentValue    - Int. Current value of clock we're on
currentIndex    - Int. Current index of currentSequence we're on
coinLength      - Length of coins array
'''


def _GetValidSequences(
    numHours,
    coins,
    counts,
    clockState,
    returnValues,
    currentSequence,
    currentValue,
    currentIndex,
    coinLength
):
    # If we have numHours coins in our sequence, we've
    # found a solution. Add it to returnValues.
    if (currentIndex == numHours):
        returnValues.append(''.join([GetCoinName(v) for v in currentSequence]))
    else:
        for i in range(coinLength):

            # If we ran out of this coin, continue
            if counts[i] == 0:
                continue

            # Determine where the next coin would go
            coinValue = coins[i]
            nextValue = (currentValue + coinValue) % numHours

            # If there's already a coin there, continue
            if clockState[nextValue]:
                continue

            # Place the coin on the clock
            currentSequence[currentIndex] = coinValue
            clockState[nextValue] = True
            counts[i] -= 1

            # Recurse
            sequences = _GetValidSequences(
                numHours,
                coins,
                counts,
                clockState,
                returnValues,
                currentSequence,
                nextValue,
                currentIndex + 1,
                coinLength
            )

            # Remove the coin from the clock to try the next coin
            clockState[nextValue] = False
            counts[i] += 1


'''
GetCoinName
Returns character that is the name of the coin based on the given value
value - value of coin to get character of
'''


def GetCoinName(value):
    if value == 1:
        return 'p'
    if value == 5:
        return 'n'
    if value == 10:
        return 'd'
    return '?'


'''
GetValidSequences
Gets list of strings representing valid clock coin sequences
numHours        - Int. Number of hours on the clock
coins           - Array of ints. Must be matched with counts. Gives value of each coin
counts          - Array of ints. Must be matched with coins. Gives count of each coin
'''


def GetValidSequences(numHours, coins, counts):
    clockState = [False] * numHours
    currentSequence = [None] * numHours

    sequences = []
    _GetValidSequences(
        numHours,
        coins,
        counts,
        clockState,
        sequences,
        currentSequence,
        0,
        0,
        len(coins)
    )

    return sequences


def main():
    numHours = 12
    coins = [1, 5, 10]
    counts = [4, 4, 4]

    start = time.time()
    for _ in itertools.repeat(None, 1000):
        GetValidSequences(numHours, coins, counts)
    end = time.time()

    # time.time() gives us difference in seconds.
    # Multiply by 1000 to get ms
    total = 1000 * (end - start)
    print(total)


main()

File: coins-on-the-clock/cpp/coins-on-the-clock.cpp
#include <string>
#include <iostream>
#include <vector>
#include <time.h>
using namespace std;

// Returns char representation of coin value
// value - value of coin to get char of
char getCoinName(int value)
{
  switch (value)
  {
  case 1:
    return 'p';
  case 5:
    return 'n';
  case 10:
    return 'd';
  case 25:
    return 'q';
  default:
    return '?';
  }
}

// Gets int arrays corresponding to valid coin sequences
// numHours             - number of hours on the clock
// values               - coin values we have (must match with counts)
// counts               - count of each coin value (must match with values)
// numValues            - number of different coin values we have
// currentSequence      - array of coin values we've placed
// clockState           - for each clock hour, true if a coin is on it
// returnValues         - reference to vector for pushing complete sequences
// currentValue         - current clock hour we're on
// currentSequenceIndex - index of currentSequence we're on
void getValidSequences(
    int numHours,
    int *values,
    int *counts,
    int numValues,
    int *currentSequence,
    bool *clockState,
    vector<string> *returnValues,
    int currentValue,
    int currentSequenceIndex)
{
  // If we're on the last hour, we have a valid sequence
  if (currentSequenceIndex == numHours)
  {

    // Copy the array and push it onto the vector
    char *newSequence = new char[numHours];
    for (int i = 0; i < numHours; i += 1)
    {
      *(newSequence + i) = getCoinName(*(currentSequence + i));
    }
    string newSequenceString = newSequence;
    returnValues->push_back(newSequenceString);
  }
  else
  {

    // For each coin value...
    for (int i = 0; i < numValues; i += 1)
    {

      // If we've run out of this coin value, continue
      if (*(counts + i) == 0)
        continue;

      // See where the next coin would go
      int value = *(values + i);
      int nextValue = (currentValue + value) % numHours;

      // If there's already a coin there, continue
      if (*(clockState + nextValue))
        continue;

      // Place the coin
      *(currentSequence + currentSequenceIndex) = value;
      *(clockState + nextValue) = true;
      *(counts + i) -= 1;

      // Recurse
      getValidSequences(
          numHours,
          values,
          counts,
          numValues,
          currentSequence,
          clockState,
          returnValues,
          nextValue,
          currentSequenceIndex + 1);

      // Remove coin
      *(clockState + nextValue) = false;
      *(counts + i) += 1;
    }
  }
}

// Gets vector of strings of valid coin sequences
// numHours  - number of hours on the clock
// values    - coin values we have (must match with counts)
// counts    - count of each coin value (must match with values)
// numValues - number of different coin values we have
vector<string> getValidSequences(
    int numHours,
    int *values,
    int *counts,
    int numValues)
{

  // Initialize clock state to all false (there are no coins after all!)
  bool *clockState = new bool[numHours]();
  int *currentSequence = new int[numHours];

  vector<string> sequences;
  getValidSequences(
      numHours,
      values,
      counts,
      numValues,
      currentSequence,
      clockState,
      &sequences,
      0,
      0);

  delete[] clockState;
  delete[] currentSequence;

  return sequences;
}

int main()
{
  int numHours = 12;
  int coins[] = {1, 5, 10};
  int counts[] = {4, 4, 4};
  int numValues = 3;

  // Get the number of ticks it takes for 1000 calculations.
  clock_t t;
  t = clock();
  for (int i = 0; i < 1000; i += 1)
  {
    getValidSequences(numHours, &coins[0], &counts[0], numValues);
  }
  t = clock() - t;

  // Divide by CLOCKS_PER_SEC to get seconds for 1000 calculations.
  // Multiple by 1000 to get ms for 1000 calculations.
  cout << ((double)t) * 1000 / CLOCKS_PER_SEC << endl;
}
